File: backend/app/jalali.py
"""تبدیل تقویم جلالی و میلادی.

بدون وابستگی خارجی نوشته شده تا نصب آفلاین ساده بماند.
الگوریتم همان jalaali استاندارد است (بر پایه شمارش روز جولیَن).
"""
from __future__ import annotations

from datetime import date
from typing import Optional

_BREAKS = [-61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210,
           1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178]

# محدوده‌ای که سال را شمسی می‌دانیم؛ خارج از آن میلادی فرض می‌شود.
JALALI_MIN_YEAR, JALALI_MAX_YEAR = 1200, 1700
GREGORIAN_MIN_YEAR, GREGORIAN_MAX_YEAR = 1800, 2400

_FA_DIGITS = str.maketrans("۰۱۲۳۴۵۶۷۸۹٠١٢٣٤٥٦٧٨٩", "01234567890123456789")


def _idiv(a: int, b: int) -> int:
    # تقسیم صحیح با قطع به سمت صفر (مانند trunc در جاوااسکریپت)
    q = abs(a) // abs(b)
    return q if (a >= 0) == (b > 0) else -q


def _imod(a: int, b: int) -> int:
    return a - _idiv(a, b) * b


def _jal_cal(jy: int) -> dict:
    bl = len(_BREAKS)
    gy = jy + 621
    leap_j = -14
    jp = _BREAKS[0]
    if jy < jp or jy >= _BREAKS[bl - 1]:
        raise ValueError(f"سال شمسی خارج از محدوده پشتیبانی: {jy}")
    jump = 0
    for i in range(1, bl):
        jm = _BREAKS[i]
        jump = jm - jp
        if jy < jm:
            break
        leap_j += _idiv(jump, 33) * 8 + _idiv(_imod(jump, 33), 4)
        jp = jm
    n = jy - jp
    leap_j += _idiv(n, 33) * 8 + _idiv(_imod(n, 33) + 3, 4)
    if _imod(jump, 33) == 4 and jump - n == 4:
        leap_j += 1
    leap_g = _idiv(gy, 4) - _idiv((_idiv(gy, 100) + 1) * 3, 4) - 150
    march = 20 + leap_j - leap_g
    if jump - n < 6:
        n = n - jump + _idiv(jump + 4, 33) * 33
    leap = _imod(_imod(n + 1, 33) - 1, 4)
    if leap == -1:
        leap = 4
    return {"leap": leap, "gy": gy, "march": march}


def _g2d(gy: int, gm: int, gd: int) -> int:
    d = (_idiv((gy + _idiv(gm - 8, 6) + 100100) * 1461, 4)
         + _idiv(153 * _imod(gm + 9, 12) + 2, 5) + gd - 34840408)
    d = d - _idiv(_idiv(gy + 100100 + _idiv(gm - 8, 6), 100) * 3, 4) + 752
    return d


def _d2g(jdn: int) -> tuple[int, int, int]:
    j = 4 * jdn + 139361631
    j = j + _idiv(_idiv(4 * jdn + 183187720, 146097) * 3, 4) * 4 - 3908
    i = _idiv(_imod(j, 1461), 4) * 5 + 308
    gd = _idiv(_imod(i, 153), 5) + 1
    gm = _imod(_idiv(i, 153), 12) + 1
    gy = _idiv(j, 1461) - 100100 + _idiv(8 - gm, 6)
    return gy, gm, gd


def _j2d(jy: int, jm: int, jd: int) -> int:
    r = _jal_cal(jy)
    return _g2d(r["gy"], 3, r["march"]) + (jm - 1) * 31 - _idiv(jm, 7) * (jm - 7) + jd - 1


def _d2j(jdn: int) -> tuple[int, int, int]:
    gy = _d2g(jdn)[0]
    jy = gy - 621
    r = _jal_cal(jy)
    jdn1f = _g2d(gy, 3, r["march"])
    k = jdn - jdn1f
    if k >= 0:
        if k <= 185:
            return jy, 1 + _idiv(k, 31), _imod(k, 31) + 1
        k -= 186
    else:
        jy -= 1
        k += 179
        if r["leap"] == 1:
            k += 1
    return jy, 7 + _idiv(k, 30), _imod(k, 30) + 1


_EPOCH_JDN = _g2d(1970, 1, 1)


def gregorian_to_jalali(d: date) -> str:
    """۲۰۲۶-۰۹-۰۶ → «۱۴۰۵/۰۶/۱۵»"""
    jy, jm, jd = _d2j(_g2d(d.year, d.month, d.day))
    return f"{jy:04d}/{jm:02d}/{jd:02d}"


def jalali_to_gregorian(jy: int, jm: int, jd: int) -> date:
    gy, gm, gd = _d2g(_j2d(jy, jm, jd))
    return date(gy, gm, gd)


def normalize_date(raw: str | date | None) -> tuple[Optional[str], Optional[date]]:
    """هر ورودی تاریخی را به جفتِ (متن شمسی، تاریخ میلادی) تبدیل می‌کند.

    خواستهٔ کارفرما: «اگر سال میلادی وارد شد شمسی شود». پس سالِ ۱۸۰۰..۲۴۰۰
    میلادی تلقی و به شمسی برگردانده می‌شود، و ۱۲۰۰..۱۷۰۰ شمسی است.
    ارقام فارسی و عربی و جداکننده‌های / - . همگی پذیرفته می‌شوند.
    مقدار نامعتبر → (None, None) تا لایه بالاتر خطای کاربرپسند بدهد.
    """
    if raw is None:
        return None, None
    if isinstance(raw, date):
        return gregorian_to_jalali(raw), raw

    text = str(raw).translate(_FA_DIGITS).strip()
    if not text:
        return None, None

    parts: list[str] = []
    current = ""
    for ch in text:
        if ch.isdigit():
            current += ch
        elif current:
            parts.append(current)
            current = ""
    if current:
        parts.append(current)
    if len(parts) < 3:
        return None, None

    try:
        y, m, d = int(parts[0]), int(parts[1]), int(parts[2])
    except ValueError:
        return None, None
    if not (1 <= m <= 12 and 1 <= d <= 31):
        return None, None

    try:
        if GREGORIAN_MIN_YEAR <= y <= GREGORIAN_MAX_YEAR:
            g = date(y, m, d)
            return gregorian_to_jalali(g), g
        if JALALI_MIN_YEAR <= y <= JALALI_MAX_YEAR:
            g = jalali_to_gregorian(y, m, d)
            return f"{y:04d}/{m:02d}/{d:02d}", g
    except (ValueError, OverflowError):
        return None, None
    return None, None


def today_jalali() -> str:
    return gregorian_to_jalali(date.today())

File: lib/jalali.js
'use strict';
/* ============================================================================
   تقویم جلالی — بدون هیچ پکیج خارجی
   ---------------------------------------------------------------------------
   الگوریتم استاندارد jalaali بر پایه شمارش روز جولیَن.
   خواسته کارفرما: اگر کاربر تاریخ میلادی وارد کند، خودکار شمسی شود.
   ========================================================================== */

const BREAKS = [-61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210,
                1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178];

/* سالِ ۱۲۰۰..۱۷۰۰ شمسی است و ۱۸۰۰..۲۴۰۰ میلادی. بین این دو بازه فاصله هست،
   پس هیچ تاریخی به دو شکل تفسیر نمی‌شود. */
const JALALI_MIN = 1200, JALALI_MAX = 1700;
const GREG_MIN   = 1800, GREG_MAX   = 2400;

const idiv = (a, b) => Math.trunc(a / b);
const imod = (a, b) => a - Math.trunc(a / b) * b;

function jalCal(jy){
  const bl = BREAKS.length;
  let gy = jy + 621, leapJ = -14, jp = BREAKS[0], jump = 0;
  if (jy < jp || jy >= BREAKS[bl - 1]) throw new RangeError('سال شمسی خارج از محدوده: ' + jy);
  for (let i = 1; i < bl; i++){
    const jm = BREAKS[i];
    jump = jm - jp;
    if (jy < jm) break;
    leapJ += idiv(jump, 33) * 8 + idiv(imod(jump, 33), 4);
    jp = jm;
  }
  let n = jy - jp;
  leapJ += idiv(n, 33) * 8 + idiv(imod(n, 33) + 3, 4);
  if (imod(jump, 33) === 4 && jump - n === 4) leapJ += 1;
  const leapG = idiv(gy, 4) - idiv((idiv(gy, 100) + 1) * 3, 4) - 150;
  const march = 20 + leapJ - leapG;
  if (jump - n < 6) n = n - jump + idiv(jump + 4, 33) * 33;
  let leap = imod(imod(n + 1, 33) - 1, 4);
  if (leap === -1) leap = 4;
  return { leap, gy, march };
}

function g2d(gy, gm, gd){
  let d = idiv((gy + idiv(gm - 8, 6) + 100100) * 1461, 4)
        + idiv(153 * imod(gm + 9, 12) + 2, 5) + gd - 34840408;
  d = d - idiv(idiv(gy + 100100 + idiv(gm - 8, 6), 100) * 3, 4) + 752;
  return d;
}

function d2g(jdn){
  let j = 4 * jdn + 139361631;
  j = j + idiv(idiv(4 * jdn + 183187720, 146097) * 3, 4) * 4 - 3908;
  const i  = idiv(imod(j, 1461), 4) * 5 + 308;
  const gd = idiv(imod(i, 153), 5) + 1;
  const gm = imod(idiv(i, 153), 12) + 1;
  const gy = idiv(j, 1461) - 100100 + idiv(8 - gm, 6);
  return { gy, gm, gd };
}

function j2d(jy, jm, jd){
  const r = jalCal(jy);
  return g2d(r.gy, 3, r.march) + (jm - 1) * 31 - idiv(jm, 7) * (jm - 7) + jd - 1;
}

function d2j(jdn){
  const gy0 = d2g(jdn).gy;
  let jy = gy0 - 621;
  const r = jalCal(jy);
  const jdn1f = g2d(gy0, 3, r.march);
  let k = jdn - jdn1f;
  if (k >= 0){
    if (k <= 185) return { jy, jm: 1 + idiv(k, 31), jd: imod(k, 31) + 1 };
    k -= 186;
  } else {
    jy -= 1; k += 179;
    if (r.leap === 1) k += 1;
  }
  return { jy, jm: 7 + idiv(k, 30), jd: imod(k, 30) + 1 };
}

const pad = n => String(n).padStart(2, '0');

/* روز مطلق برای یک تاریخ میلادی؛ مبنای همه محاسبات بازه */
function dayNumberOfDate(d){ return g2d(d.getFullYear(), d.getMonth() + 1, d.getDate()); }
function todayNumber(){ return dayNumberOfDate(new Date()); }

function fromDayNumber(n){
  const j = d2j(n);
  return `${j.jy}/${pad(j.jm)}/${pad(j.jd)}`;
}

function gregorianToJalali(d){ return fromDayNumber(dayNumberOfDate(d)); }

const FA_DIGITS = { '۰':'0','۱':'1','۲':'2','۳':'3','۴':'4','۵':'5','۶':'6','۷':'7','۸':'8','۹':'9',
                    '٠':'0','١':'1','٢':'2','٣':'3','٤':'4','٥':'5','٦':'6','٧':'7','٨':'8','٩':'9' };
const latinDigits = s => String(s).replace(/[۰-۹٠-٩]/g, c => FA_DIGITS[c]);

/**
 * هر ورودی تاریخی را به { jalali, day } تبدیل می‌کند.
 *   jalali — متن شمسی «۱۴۰۵/۰۶/۱۵» برای نمایش
 *   day    — شماره روز مطلق، برای مقایسه و محاسبه بازه
 * ورودی نامعتبر → null، تا لایه بالاتر خطای کاربرپسند بدهد.
 */
function normalize(raw){
  if (raw === null || raw === undefined) return null;
  if (raw instanceof Date && !isNaN(raw))
    return { jalali: gregorianToJalali(raw), day: dayNumberOfDate(raw) };

  const parts = latinDigits(raw).match(/\d+/g);
  if (!parts || parts.length < 3) return null;
  const y = +parts[0], m = +parts[1], d = +parts[2];
  if (!(m >= 1 && m <= 12) || !(d >= 1 && d <= 31)) return null;

  try {
    if (y >= GREG_MIN && y <= GREG_MAX){
      /* الزام کارفرما: ورودی میلادی خودکار شمسی می‌شود */
      const day = g2d(y, m, d);
      const back = d2g(day);
      if (back.gy !== y || back.gm !== m || back.gd !== d) return null;  // مثلاً ۳۱ فروردینِ میلادی
      return { jalali: fromDayNumber(day), day };
    }
    if (y >= JALALI_MIN && y <= JALALI_MAX){
      const day = j2d(y, m, d);
      const back = d2j(day);
      if (back.jy !== y || back.jm !== m || back.jd !== d) return null;  // مثلاً ۳۱ مهر
      return { jalali: `${y}/${pad(m)}/${pad(d)}`, day };
    }
  } catch (_) { return null; }
  return null;
}

module.exports = { normalize, todayNumber, fromDayNumber, gregorianToJalali, latinDigits };
